src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs

```diff
@@ -113,21 +113,24 @@ public class Frame<TContext> : Frame
                                 await FireOnCompleted();
                             }
 
-                            _application.DisposeContext(context, _applicationException);
-                        }
-
-                        // If _requestAbort is set, the connection has already been closed.
-                        if (Volatile.Read(ref _requestAborted) == 0)
-                        {
-                            ResumeStreams();
-
-                            if (_keepAlive)
+                            // If _requestAbort is set, the connection has already been closed.
+                            if (Volatile.Read(ref _requestAborted) == 0)
                             {
-                                // Finish reading the request body in case the app did not.
-                                await messageBody.Consume();
+                                ResumeStreams();
+
+                                if (_keepAlive)
+                                {
+                                    // Finish reading the request body in case the app did not.
+                                    await messageBody.Consume();
+                                }
+
+                                // ProduceEnd() must be called before _application.DisposeContext(), to ensure
+                                // HttpContext.Response.StatusCode is correct set when
+                                // IHttpContextFactory.Dispose(HttpContext) is called.
+                                await ProduceEnd();
                             }
 
-                            await ProduceEnd();
+                            _application.DisposeContext(context, _applicationException);
```

Comment:
One edge case currently not handled, is an app which calls `HttpContext.Abort()` and then throws an exception, which should probably result in `HttpContext.Response.StatusCode = 500`.  We can try to call `ProduceEnd()` even if the request has been aborted, which will correct set the status code, but I'm not sure if `ProduceEnd()` assumes the request was not aborted.